File: src/main/java/com/ryderbelserion/discord/api/options/OptionsManager.java
package com.ryderbelserion.discord.api.options;

import com.ryderbelserion.discord.api.options.interfaces.IOption;
import com.ryderbelserion.discord.api.options.types.TokenOption;
import joptsimple.OptionParser;
import joptsimple.OptionSet;
import org.jetbrains.annotations.NotNull;
import java.util.*;

public class OptionsManager {

    private final Map<String, IOption<?>> cache = new HashMap<>();

    private final OptionParser parser;

    public OptionsManager() {
        this.parser = new OptionParser();
    }

    private OptionSet options;

    public @NotNull final Optional<IOption<?>> getOptionByName(@NotNull final String name) {
        return Optional.ofNullable(this.cache.getOrDefault(name, null));
    }

    public @NotNull final Map<String, IOption<?>> getCache() {
        return Collections.unmodifiableMap(this.cache);
    }

    public @NotNull final Optional<OptionSet> getOption() {
        return Optional.of(this.options);
    }

    public void init(@NotNull final String[] args) {
        List.of(
                new TokenOption(this.parser)
        ).forEach(option -> this.cache.put(option.getName(), option.init()));

        this.options = this.parser.parse(args);
    }
}

File: src/main/kotlin/com/ryderbelserion/api/options/OptionsManager.kt
package com.ryderbelserion.api.options

import com.ryderbelserion.api.options.interfaces.IOption
import com.ryderbelserion.api.options.types.TokenOption
import joptsimple.OptionParser
import joptsimple.OptionSet
import java.util.Collections
import java.util.Optional

class OptionsManager {

    private val options = hashMapOf<String, IOption<*>>()

    private val parser: OptionParser = OptionParser()

    private var option: Optional<OptionSet> = Optional.empty()

    fun init(args: Array<String>) {
        listOf(
            TokenOption(this.parser, "Provides a discord bot token to the application.", "token")
        ).forEach { option ->
            this.options[option.name] = option.init()
        }

        this.option = Optional.of(this.parser.parse(*args))
    }

    fun getOptionByName(name: String): Optional<IOption<*>> {
        return Optional.ofNullable(this.options[name])
    }

    fun getOptions(): Map<String, IOption<*>?> {
        return Collections.unmodifiableMap(this.options)
    }

    fun getOptionSet(): Optional<OptionSet> {
        return this.option
    }
}